angular/projects/admin-nrpti/src/app/mines/mines-add-edit/mines-add-edit.component.ts

```diff
@@ -16,105 +17,197 @@ export class MinesAddEditComponent implements OnInit, OnDestroy {
-    this.loadingScreenService.setLoadingState(true, 'main');
+  /**
+   * Toggle the publish formcontrol.
+   *
+   * @param {*} event
+   * @memberof MinesAddEditComponent
+   */
+  togglePublish(event) {
+    this.myForm.controls.publish.setValue(event.checked);
+
+    this._changeDetectionRef.detectChanges();
+  }
+
+  /**
+   * Parses the form data into a mine object.
+   *
+   * @returns mine object
+   * @memberof MinesAddEditComponent
+   */
+  buildMineObject() {
     const mineItem = {};
-    mineItem['_id']                  = this.record._id;
-    mineItem['_schemaName']          = this.record._schemaName;
-    mineItem['_sourceRefId']         = this.record._sourceRefId;
-    mineItem['name']                 = this.record.name;
-    mineItem['permitNumbers']        = this.record.permitNumbers;
-    mineItem['status']               = this.record.status;
-    mineItem['type']                 = this.myForm.controls.type.value;
-    mineItem['commodities']          = this.record.commodities;
-    mineItem['tailingsImpoundments'] = this.record.tailingsImpoundments;
-    mineItem['region']               = this.record.region;
-    mineItem['location']             = this.record.location;
-    mineItem['permittee']            = this.record.permitee;
-    mineItem['summary']              = this.myForm.controls.summary.value;
-    mineItem['description']          = this.myForm.controls.description.value;
-    mineItem['links']                = this.myForm.controls.links.value.split(',');
 
-    this.loadingScreenService.setLoadingState(false, 'main');
+    mineItem['_id'] = this.mine._id;
 
-    if (!this.isEditing) {
-      // Add the mine item.
-      // Are we manually adding mines or is this import only?
-      this.router.navigate(['mines', this.record._id, 'detail']);
-    } else {
-      // Update the mine item.
-      this.factoryService.editMine(mineItem).subscribe(async res => {
-        this.router.navigate(['mines', this.record._id, 'detail']);
-      });
+    this.myForm.get('type').dirty && (mineItem['type'] = this.myForm.get('type').value);
+    this.myForm.get('description').dirty && (mineItem['description'] = this.myForm.get('description').value);
+    this.myForm.get('summary').dirty && (mineItem['summary'] = this.myForm.get('summary').value);
+    this.myForm.get('links').dirty && (mineItem['links'] = this.parseLinksFormGroups());
+
+    if (this.myForm.get('publish').dirty && this.myForm.get('publish').value) {
+      mineItem['addRole'] = 'public';
+    } else if (this.myForm.get('publish').dirty && !this.myForm.get('publish').value) {
+      mineItem['removeRole'] = 'public';
     }
+
+    return mineItem;
+  }
+
+  /**
+   * Transform the form data and save.
+   *
+   * @memberof MinesAddEditComponent
+   */
+  async submit() {
+    this.loadingScreenService.setLoadingState(true, 'main');
+
+    const mineItem = this.buildMineObject();
+
+    this.factoryService.editMine(mineItem).subscribe(async res => {
```

Comment:
Doesn't look the callback needs to be `async`.

Reply:
yeah, not sure why its async.  Probably copied from somewhere.  Will fix